Path: Added UC-CNC postprocessor script.

papaathome opened the pull request:
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

luzpaz commented:
Well done! This still needs work. I did the work for you in #4662  
What's your preference, keep this open and fix it or close in favor of #4662

papaathome commented:
Some comments on the pull request for adding an UC-CNC post processor and the pull request form.

On discussing issues with other developers:
Not done, this is a new script (based upon linuxcnc_post.py) and was developed from my own needs. I would like to share this post processor.

On rebasing:
This is my first attempt to create a pull request. In previous times I checked out other git repositories but never have used a workflow like in use for FreeCAD. The repository was branched about one hour before creating this pull request so I assume that the repository is 'rebased'.

About FreeCAD unit tests:
I cannot build FreeCAD and as such cannot do any unit testing.
Also the post processing script is new and I doubt if any unit test is touching it.
I have tested the script extensively with FreeCAD v0.17 and used is with success in v0.18 and v0.19

All commit message are well-written.
I might hope so! But English is not my native language.

About a well written pull request:
Not clear to me where I should put it, so... no?

About including issue# and fixes#:
None included, this is a new contribution.

papaathome commented:
> 
> 
> Well done! This still needs work. I did the work for you in #4662
I did see something happening in #4662 but was not paying enough attention to it.
I will give it a look now.

> What's your preference, keep this open and fix it or close in favor of #4662 ?
My preference would be to continue with this request and forget #4662 altogether.
By the way, I deleted repository 'papaathome/FreeCAD_work' and any link to it used in #4662 will be broken now.

papaathome commented:
Oops... I posted a commend in #4662 which should be posted here. Repeating...

@luzpaz
I checked the build log to see why it is failing and two things came up (for me).

The first one is simple. Around line 10986 - 11003 in the build log files are listed in the .../Path/PathScripts/port/ folder.
The script uccnc_port.py is missing.
My explanation (to be checked):
I did add the uccnc_port.py file but have not looked at the build process at all. The build process is not doing anything with the new script.

The second one is more confusing (for me) and I have no explanation for it.
At the end of the build log there is a message about a .py file that cannot be parsed.
The file in question is not near the area of ../Path/ and the chances that I by accident did hit the file is very small.

![Capture](https://user-images.githubusercontent.com/6624189/112368588-86aad080-8cdb-11eb-96e5-3ae41aeaf06e.PNG)

I do not think it is a problem related to my actions.
It would be interesting to know if the same problem also occurs in other builds but I have not checked it.
But I noticed that all builds after "#4653 opened 2 days ago by donovaly" are failing.

recap:
A simple suggestion to create a 'Pull Request' turns out to be a bit more work than adding a file and a request to take it up in the standard distribution. So far I'm enjoying the experience and I will have a look into the build process and will try to get this working.

I also will try again to be able to build FreeCAD on my machine. Last year I tried but it failed because I could not solve all dependencies for a windows based build.

Russ4262 commented:
> ...
> I also will try again to be able to build FreeCAD on my machine. Last year I tried but it failed because I could not solve all dependencies for a windows based build.

You can find great help in the forum.  Also the FC Wiki has resources for helping to compile on Windows.  Last year I made an attempt at a video to walk Windows users through the process.  Under the `Building FreeCAD` section, it is posted in the `Video Resources` section of [Building with Visual Studio 15 (2017) and 16 (2019)](https://wiki.freecadweb.org/Compile_on_Windows#Building_with_Visual_Studio_15_.282017.29_and_16_.282019.29).  I will be glad to assist as I am able.  I am on Windows 10 and building successfully with the LibPacks available.

Also, I commented over on #4662 about some code changes that need to be made for this new postprocessor due to a recent bug identified in the forum.

papaathome commented:
> > I also will try again to be able to build FreeCAD on my machine. Last year I tried but it failed because I could not solve all dependencies for a windows based build.
> 
> You can find great help in the forum. Also the FC Wiki has resources for helping to compile on Windows. Last year I made an attempt at a video to walk Windows users through the process. Under the `Building FreeCAD` section, it is posted in the `Video Resources` section of [Building with Visual Studio 15 (2017) and 16 (2019)](https://wiki.freecadweb.org/Compile_on_Windows#Building_with_Visual_Studio_15_.282017.29_and_16_.282019.29). I will be glad to assist as I am able. I am on Windows 10 and building successfully with the LibPacks available.

@Russ4262 
Thanks for the advise, I will give it a good look, because I will need it.

> Also, I commented over on #4662 about some code changes that need to be made for this new postprocessor due to a recent bug identified in the forum.

I will give that a look tomorrow. Its getting late here.  :-)

papaathome commented:
@Russ4262  
> Also, I commented over on #4662 about some code changes that need to be made for this new post processor due to a recent bug identified in the forum.

Sorry, but I can't find your comments.

I did find some remarks in the PATH/CAM group with subject "Bug Postprocessing error 'Cannot access attribute 'Proxy' of deleted object." and assume that is the bug you are talking about.
It has a fix by sliptonic » Wed Mar 24, 2021 4:09 am 
I will have to copy the changes sliptonic made to linuxcnc_post.py

papaathome commented:
recap for today:
Build now handles uccnc_post.py in a sensible way.

This pull request has 2 out of 7 items checked but I do not know enough of the work flow to get the other five done.
I've got some reading to do on this subject because I do not know how to continue in a sensible way. (Any advise is welcome!)

Travis build on linux is OK, on windows it fails.
For what I can get from the windows build log it is not a problem I introduced and I wait with actions on this point.

In the Path/CAM forum a bug was discussed and sliptonic suggested changes to linuxcnc_post.py
When his changes are pushed to the master I will have to follow it.
Not a blocking item at this moment in time. 
I have already done the changes in my local version but have not yet pushed it.

sliptonic commented on src/Mod/Path/PathScripts/post/uccnc_post.py:
```diff
@@ -0,0 +1,671 @@
+	for obj in objectslist:
+
+		# fetch machine details
+		job = PathUtils.findParentJob(obj)
```
Please check this thread.  https://forum.freecadweb.org/viewtopic.php?f=15&t=56982
This is deprecated code.  The jobs no longer have the  'MachineName' or 'MachineUnits' properties so looking up the job is worthless and will lead to the deleted proxy error. 
The fix is easy and I've made it for all the other postprocessors in this PR.  https://github.com/FreeCAD/FreeCAD/pull/4673
You should make a similar fix here before this is merged.

papaathome replied:
I have seen the topic and have your changes already in my local version. I will follow it up and push it tomorrow.

papaathome replied:
Applied and pushed the fix.

papaathome commented:
@luzpaz 
Can I ask how you did add the icon?
I looked for it but could not find out how to de it.

luzpaz commented:
@papaathome it's a label. I have triage permissions in this repo to apply it.

papaathome commented:
> 
> 
> @papaathome it's a label. I have triage permissions in this repo to apply it.

@luzpaz  And I do not have that permission and can't do it. clear.

And that is the point, in a previous message you said:
>
>
> Well done! This still needs work. I did the work for you in #4662

You did set the label in #4662 and I assumed you did it for me and started to look around how to do it.

This is my first Pull Request, ever, and clearly do not know what is expected from me. What I can do and what I must do.
I followed up on the review advise and was able to request another review.
What is next? I do not (yet) have a full understanding of the workflow and not yet found a 'step by step' guide for newbie noobs like me.

Can you advise me on what is expected from me as the next action? Enlighten me on the missing steps that I have to take to bring this to a good end. That would be very much appreciated by me.

Kind regards,
Andre.

luzpaz commented:
@papaathome I hear you :wink: So your Pull Request is getting reviewed now. You go through the review process (back and forth making revisions where/when needed) until the PR is ready to be merged.  

It's considered good form to name each commit that you push to the PR something descriptive and prepend the Workbench or suitable subject, in this case it would be `Path:` or `[Path]`. So for example: the original commit should of been named `Path: Added UC-CNC postprocessor script`.   

At the end of the PR review process, you can [squash commits](https://levelup.gitconnected.com/how-to-squash-git-commits-9a095c1bc1fc) down in to discrete summary commits. But ask the reviewer if you should. Since anytime you make a change to the PR it will trigger the Continuous Integration mechanism (which compiles FC from scratch each time).

![image](https://user-images.githubusercontent.com/4140247/112632156-21a3c780-8e0e-11eb-97f1-a071be35d477.png)

papaathome commented:
@luzpaz 
 
> 
> @papaathome I hear you...

Thanks!
...
> You go through the review process (back and forth making revisions where/when needed) until the PR is ready to be merged.

Then the main thing to do for me is watch the review(s) and respond to it.
... And in the meantime do some catch up in reading.

sliptonic commented:
Since this is your first PR, I'll give some broader feedback.  The code here looks generally fine.  Thanks for contributing!

There's only one thing specifically that I want to see changed:  Please re-tab the file using spaces instead of tab characters.  Mixing spaces and tabs is the great evil of python and makes working on the file later difficult and prone to errors.  Anytime you're working on Python, you should have your editor set to insert spaces instead of tabs.

There are some other formatting conventions that should be followed but which I'm not a stickler on. We generally follow PEP8  formatting.  That's the 'pythonic' way of doing things.

The file is over-indented.  You should use 4 space indentation.

There are other whitespace issues like trailing whitespace and missing whitespace around operators.

If you use a PEP8 delinter like Flake8  your editor can show this stuff and usually fix it automatically. 

PEP8 is also picky about line length and complains if you're over 80 columns.  I ignore this unless the lines are over 100 columns. Even then, only if it makes sense.

Same thing with block comments and aligning lists:  common sense prevails over rules.

papaathome commented:
@sliptonic 
> 
> Since this is your first PR, I'll give some broader feedback. The code here looks generally fine. Thanks for contributing!

Thanks for the kind words. :-)

> There's only one thing specifically that I want to see changed: Please re-tab the file using spaces instead of tab characters. 

Changed all tabs to 4 spaces.

> There are some other formatting conventions that should be followed but which I'm not a stickler on. We generally follow PEP8 formatting. That's the 'pythonic' way of doing things.

If I want to contribute more code to FreeCAD I better start using the rules now. I got Flake8 and ran it on the file. I tried to resolve all remarks and used a line length of 100. Still 3 lines longer than that for which I did not find an elegant way to change.

In the process I also removed a number of redundant 'global' declarations.

Pushed the new version and I will issue a new review request.
